# Homework 1: Data Cleaning and Exploratory Data Analysis 
***

**Name**: 

***

This assignment is due on Canvas by **5pm on Thursday June 13th**. Your solutions to theoretical questions should be done in Markdown directly below the associated question.  Your solutions to computational questions should include any specified Python code and results as well as written commentary on your conclusions.  Remember that you are encouraged to discuss the problems with your classmates, but **you must write all code and solutions on your own**.   

**NOTES**: 

- If you're not familiar with typesetting math directly into Markdown then by all means, do your work on paper first and then typeset it later.  Remember that there is a [reference guide](https://math.meta.stackexchange.com/questions/5020/mathjax-basic-tutorial-and-quick-reference) linked on Piazza on writing math in Markdown. 
- It is **bad form** to make your reader interpret numerical output from your code.  If a question asks you to compute some value from the data you should show your code output **AND** write a summary of the results in Markdown directly below your code. 



In [1]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt
%matplotlib inline

### Problem 1 
***

Rebecca cares about [octopuses](https://english.stackexchange.com/questions/270/what-is-the-correct-plural-of-octopus/271). She cares so much that, up and down the coast, she opens octopus rescues, octopus temporary housing, and octopus sanctuaries, specifically targeted at juveniles in need. For convenience, we will refer to all the Juvenile Octopus Rescues, Temps, and Sanctuaries as "JORTS."

Rebecca wants to estimate the average food consumption across the JORTS this month so that she can plan the food orders for next month. She has 14 Rescues, 35 Temporary houses, and 56 Sanctuaries. What an empire!

Rebecca opens up the *JORTS Manager App* on her phone, which gives her a list of all of her coastal operations. She randomly picks 15 of them, and gets ready to email their managers asking for the monthly food reports. Of course, Rebecca has taken CSCI 3022, so she knows a thing or two about sampling, and so, to get a good estimate of the monthly food consumption (kilograms per month) for typical JORTS, she intentionally chooses 2 Rescues, 5 Temporary houses, and 8 Sanctuaries.

Identify the following: 

- the population 
- the sample frame 
- the sample 
- the type of sample 
- the quantity of interest 

#### Answer:

* Population: JORTS
* Sample Frame: 14 rescues, 35 temp houses, and 56 sanctuaries
* Sample: 2 rescues, 5 temp houses, 8 sanctuaries
* Type: Stratified Sample
* Quantity: monthly food consumption in kilograms/month

### Problem 2 
***

A method to investigate the sensitivity of the sample mean and the sample median to extreme outliers and changes in the dataset is to replace one or more elements in a given dataset by a number $y$ and investigate the effect when $y$ changes. To illustrate this, consider the dataset

$$
4.6 \quad 
5.0 \quad
6.5 \quad
7.7 \quad
y \quad
4.2 \quad
1.9
$$

**Part A**: Compute the sample mean and sample median for $y=0$. Compute them both again for $y=10$.

In [11]:
#mean
num=(4.6+5+6.5+7.7+4.2+1.9)
sampleSize = 7
mean0=num/sampleSize
mean10=(num+10)/sampleSize

#median
intList = [4.6,5.0,6.5,7.7,4.2,1.9]
intList.append(0)
intList.sort()
median0= intList[int(len(intList)/2-.5)]
intList.remove(0)
intList.append(10)
median10= intList[int(len(intList)/2-.5)]


print("Mean when y=10:", mean10, "\nMean when y=0:", mean0)
print("\nMedian when y=10:", median10, "\nMedian when y=0:", median0)

Mean when y=10: 5.7 
Mean when y=0: 4.271428571428571

Median when y=10: 5.0 
Median when y=0: 4.6


**Part B**: What should $y$ be if we want the mean to be equal to $10$? What should $y$ be if we want the mean to be equal to $0$?

In [12]:
y10=(sampleSize*10-(num))
y0=-num
print("Y if mean were to be 10:", y10)
print("Y if mean were to be 0:", y0)

Y if mean were to be 10: 40.1
Y if mean were to be 0: -29.9


**Part C**: Compute the sample median for the following cases: 
- $y=10$ 
- $y=100$ 
- $y \to \infty$ 
- $y=5.01$ 
- $y=4.99$

In [13]:
print("y=10:", intList[int(len(intList)/2-.5)])
intList.remove(10)
intList.append(100)
print("y=100:", intList[int(len(intList)/2-.5)])
intList.remove(100)
intList.append(float('inf'))
print("y=inf:", intList[int(len(intList)/2-.5)])
intList.remove(float('inf'))
intList.append(5.01)
intList.sort()
print("y=5.01:", intList[int(len(intList)/2-.5)])

intList.remove(5.01)
intList.append(4.99)
intList.sort()
print("y=4.99:", intList[int(len(intList)/2-.5)])

intList.remove(4.99)



y=10: 5.0
y=100: 5.0
y=inf: 5.0
y=5.01: 5.0
y=4.99: 4.99


**Part D**: Think about the previous parts, above, and describe in words or mathematical notation the answers to the following two questions:

- By varying $y$, what is the set of all the possible values that the sample mean could take on?
- By varying $y$, what is the set of all the possible values that the sample median could take on?

#### Answer:

The mean can take on any value with any value of y. This is because y can hold any value and the sum of it with the other values can also thus take any value. Sum for new average $= y + \sum{Values}$  and this function has a range from $-\infty$ to $+\infty$ and $-\infty$ to $+\infty$ divided by the sample size, no matter what it is, still has a range of $-\infty$ to $+\infty$


However, the median can be one of three possibilities:
1. The median is 5.0 when Y is greater than 5
2. The median is 4.6 when Y is less than 4.6
3. The median is between 4.6 and 5 nad holds the value of Y when Y is is also between 4.6 and 5

### Problem 3 
***

Let $x_1, x_2, \ldots, x_n$ be $n$ observations of a variable of interest.  Recall that the sample mean $\bar{x}_n$ and sample variance $s^2_n$ are given by 

$$
\bar{x}_n = \frac{1}{n}\sum_{k=1}^n x_k \quad \textrm{and} \quad s^2_n = \frac{1}{n-1}\sum_{k=1}^n \left( x_k - \bar{x}_n\right)^2
$$

where here the subscript $n$'s indicate the number of observations in the sample. Notice that a natural computation of the variance requires two passes over the data: one to compute the mean, and a second to subtract the mean from each observation and compute the sum of squares. It is often useful to be able to compute the variance in a single pass, inspecting each value $x_k$ only once; for example, when the data are being collected without enough storage to keep all the values, or when costs of memory access dominate those of computation. In this problem you will explore two methods for such an _online_ computation of the mean and variance.  

**Part A**: Show algebraically that the following relation holds between the mean of the first $n-1$ observations and the mean of all $n$ observations: 

$$
\bar{x}_n = \bar{x}_{n-1} + \frac{x_n - \bar{x}_{n-1}}{n}
$$

**Part B**: Write a function `my_sample_mean` that takes as its input a numpy array and returns the mean of that numpy array using the formulas from class (written above). You may *not* use numpy's built in mean function. Write another function `my_sample_var` that takes as its input a numpy array and returns the variance of that numpy array, again using the formulas from class (written above). You may *not* use any built-in sample variance functions.

In [14]:
myA = np.array([1,100,100,-100])

def my_sample_mean(A):
    total = 0
    sampleSize = len(A)
    for x in np.nditer(A):
        total = total + x
    return total/sampleSize

def my_sample_var(A):
    return np.amax(A)-np.amin(A)
    
print("Tester sample Mean:", my_sample_mean(myA), "\nTester sample variance:",my_sample_var(myA))

Tester sample Mean: 25.25 
Tester sample variance: 200


**Part C**: Use your functions from Part B to compute the sample mean and sample variance of the following array, which contains the counts of perfectly round suckers found on a set of aquarium octopuses.

`octopus_suckers = [25, 29, 40, 19, 7, 6, 3, 11, 19, 21, 22, 45, 27]`

In [15]:
Oct = np.array([25, 29, 40, 19, 7, 6, 3, 11, 19, 21, 22, 45, 27])
OctMean = my_sample_mean(Oct)
print("Mean:", OctMean)
print("Variance:", my_sample_var(Oct))

Mean: 21.076923076923077
Variance: 42


**Part D**: Implement a third function called `update_mean` that implements the formula whose valdity you proved in Part A. (Note: this function will need to take as its input three things: $x_n$, $\bar{x}_{n-1}$ and $n$.)

Use this function to compute the values that you get from taking the mean of the first suckers count, the first two suckers counts, the first three suckers counts, and so on up to all the suckers counts. Store your means in a numpy array called `sucker_means`.

In [29]:
sucker_mean = []

def update_mean(xn, xn1, n):
    x = xn1 + ((xn-xn1)/n)
    return(x)

prevMean = Oct[0]

for n in range(1, Oct.size):
    
    prevMean = update_mean(Oct[n], prevMean, n)
    sucker_mean.append(prevMean)
    
    
np.asarray(sucker_mean)
sucker_mean = np.array(sucker_mean)
print(sucker_mean)    

    

[29.         34.5        29.33333333 23.75       20.2        17.33333333
 16.42857143 16.75       17.22222222 17.7        20.18181818 20.75      ]


### Problem 4
*** 
Some claim that the final hours aboard the RMS Titanic were marked by "class warfare" in which the people with first-class tickets took all the good spots on the lifeboats; others claim that the final hours were characterized by male chivalry, in which the men valiantly gave up their positions in the boats and succumbed bravely to the depths of the Atlantic. 

We have the data on survival rates by class and by sex, so let's figure out whether there is evidence for these scenarios. Access the titanic data in `titanic_data.csv` and store it in a Pandas DataFrame. The data contains information pertaining to class status (**Pclass**), survival of passengers (**Survived**), and gender (**Sex**), among others.  

In [33]:
df = pd.read_csv("DataHW/titanic_data.csv")
df.head(15)

PassengerId  Survived  Pclass  \
0             1         0       3   
1             2         1       1   
2             3         1       3   
3             4         1       1   
4             5         0       3   
5             6         0       3   
6             7         0       1   
7             8         0       3   
8             9         1       3   
9            10         1       2   
10           11         1       3   
11           12         0       1   
12           13         0       3   
13           14         0       3   
14           15         0       3   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  36.0      1   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  18.0      1   
2                              Heikkinen, Miss. Laina  female  14.0      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  27.0      1   
4                            Allen, Mr. William Henry    male  63.0      0   
5                                    Moran, Mr. James    male  14.0      0   
6                             McCarthy, Mr. Timothy J    male  39.0      0   
7                      Palsson, Master. Gosta Leonard    male   NaN      3   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female   NaN      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female   NaN      1   
10                    Sandstrom, Miss. Marguerite Rut  female  42.0      1   
11                           Bonnell, Miss. Elizabeth  female  30.0      0   
12                     Saundercock, Mr. William Henry    male   NaN      0   
13                        Andersson, Mr. Anders Johan    male  43.0      1   
14               Vestrom, Miss. Hulda Amanda Adolfina  female   NaN      0   

    Parch            Ticket     Fare Cabin Embarked  
0       0         A/5 21171   7.2500   NaN        S  
1       0          PC 17599  71.2833   C85        C  
2       0  STON/O2. 3101282   7.9250   NaN        S  
3       0            113803  53.1000  C123        S  
4       0            373450   8.0500   NaN        S  
5       0            330877   8.4583   NaN        Q  
6       0             17463  51.8625   E46        S  
7       1            349909  21.0750   NaN        S  
8       2            347742  11.1333   NaN        S  
9       0            237736  30.0708   NaN        C  
10      1           PP 9549  16.7000    G6        S  
11      0            113783  26.5500  C103        S  
12      0         A/5. 2151   8.0500   NaN        S  
13      5            347082  31.2750   NaN        S  
14      0            350406   7.8542   NaN        S

**Part A**: Consider the two claims: class warfare, and male chivalry. Suppose that class warfare occurred in the final hours aboard the Titanic.  What patterns might you expect to see in the data?  Suppose that male chivalry was widespread during the final hours insteas. What patterns might you then expect to see in the data?  Explain both of these hypothesized patterns in words. Are these two hypotheses mutually exclusive or not?

#### Answer:
If class warfare occured, we would expect there to be a signifigant difference between survival rates of seperate classes. If male chivaly were widespread, then women would have a signifigantly higher survival rate.

These cases are not mutually exclusive and can both have occured. The two hypothesis are as follows:
1. Class warfare occured as the titanic was sinking and affected survival
2. Male chivaly occured as the titanic was sinking and affected survival 

I will analyze the data and the survival rates of groups to make wider claims about class warfare and male chivaly on the titanic.

**Part B**: Create a clean data set by removing any rows from the DataFrame that are missing values corresponding to **Survived**, **Pclass**, **Age**, or **Sex**. Store the clean data in a DataFrame called dfTitanic. 

In [34]:
dfTitanic = df.copy(deep=True)

dfTitanic.dropna(subset=["Survived","Pclass","Age", "Sex"])

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
5              6         0       3   
6              7         0       1   
10            11         1       3   
11            12         0       1   
13            14         0       3   
15            16         1       2   
16            17         0       3   
17            18         1       2   
18            19         0       3   
19            20         1       3   
20            21         0       2   
21            22         1       2   
22            23         0       3   
23            24         1       1   
26            27         0       3   
28            29         1       3   
29            30         0       3   
30            31         0       1   
32            33         1       3   
33            34         0       2   
34            35         0       1   
35            36         0       1   
36            37         1       3   
37            38         0       3   
39            40         1       3   
..           ...       ...     ...   
858          859         1       3   
859          860         0       3   
860          861         0       3   
861          862         0       2   
862          863         1       1   
863          864         0       3   
864          865         0       2   
865          866         1       2   
866          867         1       2   
867          868         0       1   
868          869         0       3   
869          870         1       3   
870          871         0       3   
871          872         1       1   
873          874         0       3   
874          875         1       2   
875          876         1       3   
876          877         0       3   
877          878         0       3   
878          879         0       3   
879          880         1       1   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  36.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  18.0      1   
2                               Heikkinen, Miss. Laina  female  14.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  27.0      1   
4                             Allen, Mr. William Henry    male  63.0      0   
5                                     Moran, Mr. James    male  14.0      0   
6                              McCarthy, Mr. Timothy J    male  39.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female  42.0      1   
11                            Bonnell, Miss. Elizabeth  female  30.0      0   
13                         Andersson, Mr. Anders Johan    male  43.0      1   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  19.0      0   
16                                Rice, Master. Eugene    male  23.0      4   
17                        Williams, Mr. Charles Eugene    male  45.0      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  16.0      1   
19                             Masselmani, Mrs. Fatima  female  29.0      0   
20                                Fynney, Mr. Joseph J    male  35.0      0   
21                               Beesley, Mr. Lawrence    male  28.0      0   
22                         McGowan, Miss. Anna "Annie"  female  24.0      0   
23                        Sloper, Mr. William Thompson    male  58.0      0   
26                             Emir, Mr. Farred Chehab    male  22.0

**Part C**: Determine the fraction of survivors from each passenger class. 

In [35]:
class1Surivors = sum(dfTitanic[dfTitanic["Pclass"] == 1]["Survived"])
class2Survivors = sum(dfTitanic[dfTitanic["Pclass"] == 2]["Survived"])
class3Survivors = sum(dfTitanic[dfTitanic["Pclass"] == 3]["Survived"])

class1 = sum(dfTitanic["Pclass"] == 1)
class2 = sum(dfTitanic["Pclass"] == 2)
class3 = sum(dfTitanic["Pclass"] == 3)

rat1 = class1Surivors/class1
rat2 = class2Survivors/class2
rat3 = class3Survivors/class3

print("First Class:", rat1, "\nSecond Class:",rat2, "\nThird Class",rat3)

First Class: 0.6342592592592593 
Second Class: 0.4891304347826087 
Third Class 0.2505091649694501


**Part D**: Compute the fraction of survivors according to class and gender.  Did men in first class or women in third class have a higher survival rate? 

In [36]:
menClass1 = sum(dfTitanic[dfTitanic["Pclass"] == 1]["Survived"])

**Part E**: How would you characterize the distribution of **AGE**? (By _characterize_ we mean that you should indicate whether the data are unimodal, bimodal, multimodal, symmetric, negatively skewed, positively skewed, etc.)  Make any necessary graphical summaries to justify your conclusion. 

**Part F**: Were the median and mean ages for females who survived higher or lower than for females who did not survive?  

**Part G**: Do the data suggest class warfare, male chivalry, or some combination of both characteristics in the final hours aboard the Titanic?  Justify your conclusion based on the computations done above, or do any other analysis that you like, but be sure to clearly justify your conclusion. 

### Problem 5 - Dirty Data
***
Access the data from url https://www.stat.berkeley.edu/~statlabs/data/babies.data and store the information in a Pandas DataFrame.  A description of the variables can be found at https://www.stat.berkeley.edu/~statlabs/labs.html.  These data are a subset from a much larger study dealing with child health and development. 


In [4]:
df = pd.read_csv("https://www.stat.berkeley.edu/~statlabs/data/babies.data", delim_whitespace=True)
df.head(10)

bwt  gestation  parity  age  height  weight  smoke
0  120        284       0   27      62     100      0
1  113        282       0   33      64     135      0
2  128        279       0   28      64     115      1
3  123        999       0   36      69     190      0
4  108        282       0   23      67     125      1
5  136        286       0   25      62      93      0
6  138        244       0   33      62     178      0
7  132        245       0   23      65     140      0
8  120        289       0   25      62     125      0
9  143        299       0   30      66     136      1


**Part A**: Create a clean data set that removes subjects if any observations on the subject are unknown.  Note that that collectors of the data set used values like $9$, $99$, $999$, to denote unknown values.  You can look at the documentation linked in the problem description to determine which unknown-value marker was used for each characteristics.  Store the modified data set in a Pandas DataFrame called dfBabies. 

**Part B**: Use the information in dfBabies to create a density histogram of the birth weights of babies whose mothers have never smoked (smoke=0) and another histogram placed directly below the first in the same graphics device for the birth weights of babies whose mothers currently smoke (smoke=1).  Make the range of the horizontal axis $30$ to $180$ (ounces) for both histograms.  Make sure to give each subplot titles and label axes. 

**Part C**: Based on the histograms in **Part B**, characterize the distribution of baby birth weights for both non-smoking and smoking mothers. 

**Part D**: What are the mean and median weight difference between babies of smokers and non-smokers?  Can you think of any reason not to use the mean as a measure of center to compare birth weights for this problem? 

**Part E**: Create side-by-side box-and-whisker plots to compare the birth weights of babies whose mothers never smoked and those who currently smoke.  Use the box-and-whisker plot conventions discussed in lecture. 

**Part F**: Using the box-and-whisker plots from **Part E** comment on the distributions of body weights of babies within each smoking / non-smoking groups as well as the comparison of the two classes. 